### PR TITLE
Optimize DeepSeek V4 MoE routing performance

### PR DESCRIPTION
**Pull request by zhaozhaozz**
## Summary

- add MoE profiling/export flags for L2 swimlane, PMU, and Kernel Insight collection
- reduce MoE dispatch/combine reshape and metadata overhead
- replace the router gate dot vector reduction path with cube `pl.matmul` / `pl.matmul_acc` accumulation
- address review feedback by making gate tiling validation unconditional and honoring explicit PTOAS/PTO ISA environment overrides

## Results

- Remote NPU strict validation passes for `python models/deepseek/v4/moe_router.py -p a2a3`; `x_norm`, `indices`, and `weights` all pass without relaxed comparators.
- Router-only profiling with L2 swimlane and PMU enabled improves Total Test Time from `1170.34 us` to `482.52 us` (`-58.8%`).
- The gate path exec time improves from `784.52 us` to `34.58 us` (`-95.6%`) after switching the gate dot to the cube matmul path.

## Validation

- `python3 -m py_compile models/deepseek/v4/moe_router.py models/deepseek/v4/moe_dispatch.py models/deepseek/v4/moe_combine.py tools/export_all_kernel_insight.py`
- `uv run ruff check models/deepseek/v4/moe_router.py models/deepseek/v4/moe_dispatch.py models/deepseek/v4/moe_combine.py tools/export_all_kernel_insight.py`
- remote NPU: `python models/deepseek/v4/moe_router.py -p a2a3`
- remote NPU profiling: `python models/deepseek/v4/moe_router.py -p a2a3 --enable-l2-swimlane --enable-pmu 2`

## Notes

- The current source does not add a relaxed BF16 comparator; the BF16 1-ULP check was only used as a diagnostic for boundary samples.
- `gate_dot` is a cube kernel, so Kernel Insight export may require a cube target override with older case-generation scripts.
- The reviewed `pl.full` suggestions for route metadata initialization were tested but not retained because they currently trigger PyPTO/PTOAS type conflicts when mixed with dynamic scalar reads/writes.

### CODE REVIEW

File: models/deepseek/v4/moe_router.py [hunk 1]
```diff
@@ -29,6 +29,12 @@
 # tiling
 D_CHUNK          = 128 if T >= 128 else (256 if T >= 64 else 512)
 D_BLOCKS         = D // D_CHUNK
+GATE_D_CHUNK     = 512
+GATE_D_BLOCKS    = D // GATE_D_CHUNK
+GATE_N_CHUNK     = N_EXPERTS if N_EXPERTS <= 32 else 16
+GATE_N_BLOCKS    = N_EXPERTS // GATE_N_CHUNK
+assert D % GATE_D_CHUNK == 0
+assert N_EXPERTS % GATE_N_CHUNK == 0
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/moe_router.py]: Handled in `f419739d`: replaced the gate tiling `assert`s with unconditional `ValueError` checks before deriving the block counts. Local static checks and remote NPU router strict validation passed.

File: tools/export_all_kernel_insight.py [hunk 2]
```diff
@@ -47,13 +47,17 @@
...
 def default_ptoas_root() -> Path:
-    local = REPO_ROOT / "PTOAS"
-    if (local / "test/npu_validation/scripts/generate_testcase.py").is_file():
-        return local
-    return repo_path(os.environ.get("PTOAS_ROOT", local))
+    for local in (REPO_ROOT / "PTOAS", REPO_ROOT.parent / "PTOAS"):
+        if (local / "test/npu_validation/scripts/generate_testcase.py").is_file():
+            return local
+    fallback = REPO_ROOT / "PTOAS"
+    return repo_path(os.environ.get("PTOAS_ROOT", fallback))
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply tools/export_all_kernel_insight.py]: Handled in `f419739d`: `PTOAS_ROOT` and `PTO_ISA_ROOT` are now checked before local auto-discovery, and valid explicit roots are honored first. Also covered by a local env override smoke check.

File: models/deepseek/v4/moe_combine.py [hunk 3]
```diff
@@ -61,29 +58,30 @@ def moe_combine(
                 routed_y_buf[r0 : r0 + N_LOCAL_EXPERTS, d0 : d0 + COL_CHUNK] = pl.full(
                     [N_LOCAL_EXPERTS, COL_CHUNK], dtype=pl.BF16, value=0.0
                 )
-        for r in pl.range(T * N_LOCAL_EXPERTS):
-            pl.write(routed_w_buf, [r], 0.0)
+        for t in pl.range(T):
+            for e in pl.range(N_LOCAL_EXPERTS):
+                pl.write(routed_w_buf, [t, e], 0.0)
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/moe_combine.py:63]: Tried this, but did not retain it. The current PyPTO/PTOAS path still mixes tensor-view and scalar pointer usage around dynamically indexed route metadata in the full MoE pipeline, so the scalar-write initialization remains the safer form for this PR.

File: models/deepseek/v4/moe_dispatch.py [hunk 4]
```diff
@@ -77,50 +77,34 @@ def moe_dispatch(
...
-            pl.write(recv_scale_dq_flat, [r], 0.0)
-            pl.write(recv_weights_flat, [r], 0.0)
-            pl.write(recv_token_flat, [r], pl.cast(0, pl.INT32))
         for e in pl.range(N_LOCAL_EXPERTS):
-            pl.write(count_flat, [e], pl.cast(0, pl.INT32))
+            for s in pl.range(RECV_MAX):
+                pl.write(recv_scale_dq, [e, s], 0.0)
+                pl.write(recv_weights, [e, s], 0.0)
+                pl.write(recv_token, [e, s], pl.cast(0, pl.INT32))
+            pl.write(recv_expert_count, [e, 0], pl.cast(0, pl.INT32))
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/moe_dispatch.py:94]: Tried this, but did not retain it. In the full MoE compile, bulk `pl.full` initialization of metadata tensors conflicts with later dynamic scalar reads/writes of the same buffers, producing a PTOAS type mismatch between tensor-view and pointer forms. Kept the existing scalar initialization.

File: models/deepseek/v4/moe_router.py [hunk 5]
```diff
@@ -84,49 +90,61 @@ def moe_router(
...
-                w_row = gate_w[expert_i : expert_i + 1, d0 : d0 + D_CHUNK]
-                prod = pl.col_expand_mul(x_chunk, w_row)
-                score_acc = pl.add(
-                    score_acc,
-                    pl.reshape(pl.row_sum(prod), [1, T]),
-                )
-            score_acc_buf[:, :] = score_acc
-            bias = pl.read(gate_bias, [expert_i])
-            logits = pl.load(score_acc_buf, [0, 0], [1, T])
+        for expert0 in pl.range(0, N_EXPERTS, GATE_N_CHUNK):
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/moe_router.py:137]: Not changed in this PR. This is a separate parallelization strategy change for `gate_score_post`; it needs dedicated semantic and performance validation because it changes how multiple chunks store into `route_scores` and `biased_scores`.

File: models/deepseek/v4/moe_router.py [hunk 6]
```diff
@@ -84,49 +90,61 @@ def moe_router(
...
+            logits = pl.load(gate_logits, [0, expert0], [T, GATE_N_CHUNK])
             zero = pl.mul(logits, 0.0)
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/moe_router.py:139]: Tried this, but did not retain it. In this context `logits` is a Tile from `pl.load`, while `pl.full([T, GATE_N_CHUNK], ...)` creates a Tensor, so replacing the expression directly causes Tensor/Tile mixing during compile.

File: models/deepseek/v4/moe_router.py [hunk 7]
```diff
@@ -84,49 +90,61 @@ def moe_router(
...
-            for t in pl.unroll(T):
-                pl.write(route_flat, [t * SCORE_PAD + expert_i], pl.read(score_row_flat, [t]))
-                pl.write(biased_flat, [t * SCORE_PAD + expert_i], pl.read(biased_row_flat, [t]))
-    route_scores = pl.reshape(route_flat, [T, SCORE_PAD])
-    biased_scores = pl.reshape(biased_flat, [T, SCORE_PAD])
+            bias_row = pl.reshape(
+                pl.load(gate_bias, [expert0], [GATE_N_CHUNK]),
+                [1, GATE_N_CHUNK],
+            )
+            bias_tile = pl.col_expand_mul(pl.add(zero, 1.0), bias_row)
```
[the remark this answers is not in the document]
Comment (zhaozhaozz) [reply models/deepseek/v4/moe_router.py:148]: Tried this, but did not retain it for the same Tile/Tensor reason as the zero-tile suggestion: the existing expression derives the tile shape/type from `logits`, while direct `pl.full` creates a Tensor in this context.
